stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -112,19 +114,30 @@ internal class ChannelControllerImpl(
     val unfilteredMessages = _messages.map { it.values.toList() }
 
     /** a list of messages sorted by message.createdAt */
-    override val messages: LiveData<List<Message>> = messagesTransformation(_messages)
+    val sortedVisibleMessages = messagesTransformation(_messages).stateIn(domainImpl.scope, SharingStarted.Eagerly, emptyList())
+    override val messages: LiveData<List<Message>> = sortedVisibleMessages.asLiveData()
+
+    private val _messagesState: StateFlow<ChannelController.MessagesState> =
+        _loading.combine(sortedVisibleMessages) { loading: Boolean, messages: List<Message> ->
+            when {
+                loading -> ChannelController.MessagesState.Loading
+                messages.isEmpty() -> ChannelController.MessagesState.OfflineNoResults
+                else -> ChannelController.MessagesState.Result(messages)
+            }
+        }.stateIn(domainImpl.scope, SharingStarted.Eagerly, ChannelController.MessagesState.NoQueryActive)
```

Comment:
Good to discuss this since the design/UI side of things can be different.

- some UIs show the loading icon overlayed on the chat messages. So showing it while there are results works well.
- other UIs only show the loading icon when there are no results available yet.. 

If i look at our designs it looks like we're trying to do the 2nd, this implement of state is different than those 2 options. (it triggers the loading interface once for offline and once for online)

Reply:
Good point, we can split loading for two cases just `Loding` and `LoadingWithCached(val cachedResult: Collection<Message>)`

Reply:
This flow is following the 2nd option.
Loading is only returned when the first query is been done, doesn't mind if it is online or offline (both are done in parallel)
We have another livedata that represent that there is another query in process, used to show the "loading more" progressbar

Reply:
I feel like I'm obligated to plug [my own content](https://zsmb.co/designing-and-working-with-single-view-states-on-android/) about distinct vs non-distinct states, as that's exactly what this discussion is about.

TLDR: There's nothing wrong with having a regular class with both a `loading` and `results` property at the same time, if those things are present at the same time on the UI. Not everything is neatly modellable by distinct cases of sealed classes, however awesome those are.